devito/passes/clusters/cse.py

```diff
@@ -92,61 +126,98 @@ def _cse(maybe_exprs, make, min_cost=1, mode='default'):
+        # Start with the largest
+        cost = key(max(candidates, key=key))
+        if cost < min_cost:
+            break
+        candidates = [c for c in candidates if c.cost == cost]
 
         # Apply replacements
-        # The extracted temporaries are inserted before the first expression
-        # that contains it
-        scheduled = []
-        updated = []
-        for e in processed:
-            pe = e
-            for k, v in chosen:
-                if not k.conditionals == e.conditionals:
-                    continue
-                pe, changed = _uxreplace(pe, {k.expr: v})
-                if changed and v not in scheduled:
-                    updated.append(pe.func(v, k.expr, operation=None))
-                    scheduled.append(v)
-            updated.append(pe)
-        processed = updated
-
-        # Update `exclude` for the same reasons as above -- to rule out CSE across
-        # Dimension-independent data dependences
-        exclude.update(scheduled)
-
-    # At this point we may have useless temporaries (e.g., r0=r1). Let's drop them
-    processed = _compact_temporaries(processed, exclude)
+        chosen = [(c, scheduled.get(c.key) or make()) for c in candidates]
+        exprs = _inject(exprs, chosen, scheduled)
+
+    # Drop useless temporaries (e.g., r0=r1)
+    processed = _compact(exprs, exclude)
+
+    # Ensure topo-sorting ('basic' doesn't require it)
+    if mode in ('smartsort', 'advanced'):
+        processed = _toposort(processed)
 
     return processed
 
 
-def _compact_temporaries(exprs, exclude):
+def _inject(exprs, chosen, scheduled):
     """
-    Drop temporaries consisting of isolated symbols.
+    Insert temporaries into the expression list.
+
+    The resulting expression list may not be topologically sorted. The caller
+    is responsible for ensuring that.
+    """
+    processed = []
+    for e in exprs:
+        pe = e
+        for k, v in chosen:
+            if k.conditionals != e.conditionals:
+                continue
+
+            if e.lhs is v:
+                # This happens when `k.expr` wasn't substituted in a previous
+                # iteration because `k.sources` (whose construction
+                # is based on heuristics to avoid a combinatorial explosion)
+                # didn't include all of the `k.expr` occurrences across `exprs`,
+                # in particular those as part of a middle-term in a n-ary operation
+                # (e.g., `b*c` in `a*b*c*d`)
+                assert k.expr == e.rhs
+                continue
+
+            subs = k.as_subs(v)
+
+            pe, changed = _uxreplace(pe, subs)
+
+            if changed and k.key not in scheduled:
+                processed.append(pe.func(v, k.expr, operation=None))
+                scheduled[k.key] = v
+
+        processed.append(pe)
+
+    return processed
+
+
+def _compact(exprs, exclude):
+    """
+    Drop useless temporaries:
```

Comment:
Are there also edge cases where it creates expressions like `r1 = r0`? I'm sure we had some issue like this

Reply:
One of the roles of `_compact` is catching them all !